src/com/cloudogu/ces/cesbuildlib/Git.groovy

```diff
@@ -309,28 +365,69 @@ class Git implements Serializable {
      * configured max retries before failing.
      *
      * @param args git arguments
+     * @return Returns an array with a string array of two elements.
+     *         The first element contains the command out put. The second element contans the command status code
      */
-    protected void executeGitWithCredentials(String args) {
+    protected CommandOutput executeGitWithCredentials(String args) {
         if (credentials) {
             script.withCredentials([script.usernamePassword(credentialsId: credentials,
                     passwordVariable: 'GIT_AUTH_PSW', usernameVariable: 'GIT_AUTH_USR')]) {
                 def pushResultCode = 128
+                def commandOutput = ""
                 def retryCount = 0
                 while (pushResultCode == 128 && retryCount < maxRetries) {
                     if (retryCount > 0) {
                         script.echo "Got error code ${pushResultCode} - retrying in ${retryTimeout} ms ..."
                         sleep(retryTimeout)
                     }
                     ++retryCount
-                    pushResultCode = script.sh returnStatus: true, script: "git -c credential.helper=\"!f() { echo username='\$GIT_AUTH_USR'; echo password='\$GIT_AUTH_PSW'; }; f\" ${args}"
-                    pushResultCode = pushResultCode as int
+                    def out = executeGit("-c credential.helper=\"!f() { echo username='\$GIT_AUTH_USR'; echo password='\$GIT_AUTH_PSW'; }; f\" ${args}")
+                    pushResultCode = out.exitCode as int
+                    commandOutput = out.stdout
                 }
                 if (pushResultCode != 0) {
                     script.error "Unable to execute git call. Retried ${retryCount} times. Last error code: ${pushResultCode}"
                 }
+                return new CommandOutput(commandOutput, pushResultCode)
             }
         } else {
-            script.sh "git ${args}"
+            return executeGit(args)
+        }
+    }
+
+    /**
+     * Executes a git command.
+     *
+     * @param args git arguments
+     * @return Returns an array with a string array of two elements.
+     *         The first element contains the command out put. The second element contans the command status code
+     */
+    protected CommandOutput executeGit(String args) {
+        try {
+            def status = script.sh(
+                    script: "git ${args} > output",
+                    returnStatus: true
+            )
+            return new CommandOutput(sh.returnStdOut("cat output"), (status) ? status as int : 0)
+        } finally {
+            script.sh "rm -f output"
+        }
+    }
```

Comment:
Is there any other way to get stdout and exit code? 
This seems not to be a good practice but is the only way in my opinion.

Reply:
I don't know a better way to get both return code and stdout.
But: Actually we only need either stdout or return code for our individual use cases.
The commands that need a retry (push and pull) need the return code.
Other commands need the stdout (e.g. getTag).

So we can provide different methods for each use case that either return stdout or return code. This way we work around having to return both from one method call.
In the end, this will reduce complexity because we don't have to handle piping to a file and deleting the file afterwards. This would also cause potential concurrency issues, because the file was always named `output`.